File: src/javaCode/boj/softeer/SafeBreaking.java
package javaCode.boj.softeer;

import org.jetbrains.annotations.NotNull;

import java.io.*;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class SafeBreaking {

    private static BufferedWriter bw;
    private static BufferedReader br;
    private static StringTokenizer st;

    private static int N, W;

    private static class Bag implements Comparable<Bag> {
        int m;
        int p;

        public Bag(int m, int p) {
            this.m = m;
            this.p = p;
        }

        @Override
        public int compareTo(@NotNull Bag o) {
            return o.p - this.p;
        }
    }

    public static void main(String args[]) throws IOException {
        bw = new BufferedWriter(new OutputStreamWriter(System.out));
        br = new BufferedReader(new InputStreamReader(System.in));

        st = new StringTokenizer(br.readLine());
        W = Integer.parseInt(st.nextToken());
        N = Integer.parseInt(st.nextToken());

        PriorityQueue<Bag> bags = new PriorityQueue<Bag>(N);

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            int m = Integer.parseInt(st.nextToken());
            int p = Integer.parseInt(st.nextToken());

            bags.offer(new Bag(m, p));
        }

        long result = 0;
        int max = W;
        while (!bags.isEmpty()) {
            Bag bag = bags.poll();
            if (bag.m > max) {
                result += (long) max * bag.p;
                break;
            }
            result += (long) bag.m * bag.p;
            max -= bag.m;
        }
        System.out.print(result);
    }
}

File: src/kotlinCode/softeer/SafeBreaking.kt
package kotlinCode.softeer

import java.util.PriorityQueue

private val br = System.`in`.bufferedReader()
private val bw = System.out.bufferedWriter()

data class Bag(val m: Int, val p: Int)

fun main() = with(br) {
    val (w, n) = readLine().split(" ").map { it.toInt() }
    val bags = PriorityQueue<Bag> { a, b ->
        if (a.p > b.p) {
            -1
        } else if (a.p == b.p) {
            b.m - a.m
        } else {
            1
        }
    }

    repeat(n) {
        val (m, p) = readLine().split(" ").map { it.toInt() }
        bags.add(Bag(m, p))
    }

    println(bags.joinToString(" "))

    var result = 0L
    var max = w
    while (bags.isNotEmpty()) {
        val bag = bags.poll()
        if (bag.m >= max) {
            result += max * bag.p
            break
        }
        result += bag.m * bag.p
        max -= bag.m
    }
    println(result)
}